clients/src/test/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumerTest.java

```diff
@@ -574,6 +581,148 @@ public void testPollLongThrowsException() {
+        verify(applicationEventHandler).add(any(FetchCommittedOffsetsEvent.class));
+        CompletableApplicationEvent<Map<TopicPartition, OffsetAndMetadata>> event1 = getLastEnqueuedEvent();
+        assertThrows(TimeoutException.class, () -> ConsumerUtils.getResult(event1.future(), time.timer(timeoutMs)));
+        assertTrue(consumer.hasPendingOffsetFetchEvent());
+
+        clearInvocations(applicationEventHandler);
+
+        // Sleep past the event's expiration, causing the poll() to *not* reuse the pending fetch. A new event
+        // is created and added to the application event queue.
+        time.sleep(event1.deadlineMs() - time.milliseconds());
+        consumer.poll(Duration.ofMillis(timeoutMs));
+        verify(applicationEventHandler).add(any(FetchCommittedOffsetsEvent.class));
+        CompletableApplicationEvent<Map<TopicPartition, OffsetAndMetadata>> event2 = getLastEnqueuedEvent();
+        assertNotEquals(event1, event2);
+        assertThrows(TimeoutException.class, () -> ConsumerUtils.getResult(event2.future(), time.timer(timeoutMs)));
+        assertTrue(consumer.hasPendingOffsetFetchEvent());
+    }
+
+    @ParameterizedTest
+    @MethodSource("offsetFetchExceptionSupplier")
+    public void testOffsetFetchPendingEventErrors(Throwable reportedFailure,
+                                                  boolean expectErrorFromPoll,
+                                                  boolean expectPending) {
+        // Interrupt the thread and then clear it so that the interrupt flag is in a known good state before starting.
+        try {
+            Thread.currentThread().interrupt();
+        } finally {
+            assertTrue(Thread.interrupted());
+        }
+
+        consumer = newConsumer();
+        long timeoutMs = 0;
+        doReturn(Fetch.empty()).when(fetchCollector).collectFetch(any(FetchBuffer.class));
+        consumer.assign(Collections.singleton(new TopicPartition("topic1", 0)));
+
+        // Set up a pending offset fetch event.
+        consumer.poll(Duration.ofMillis(timeoutMs));
+        verify(applicationEventHandler).add(any(FetchCommittedOffsetsEvent.class));
+        CompletableApplicationEvent<Map<TopicPartition, OffsetAndMetadata>> event1 = getLastEnqueuedEvent();
+        assertThrows(TimeoutException.class, () -> ConsumerUtils.getResult(event1.future(), time.timer(timeoutMs)));
+        assertTrue(consumer.hasPendingOffsetFetchEvent());
+
+        clearInvocations(applicationEventHandler);
+
+        assertNull(consumerUtilsMockedStatic);
+        consumerUtilsMockedStatic = mockStatic(ConsumerUtils.class);
+        consumerUtilsMockedStatic.when(() -> ConsumerUtils.getResult(any(Future.class), any(Timer.class))).thenThrow(reportedFailure);
+        consumerUtilsMockedStatic.when(() -> ConsumerUtils.getResult(any(Future.class))).thenThrow(reportedFailure);
+        consumerUtilsMockedStatic.when(() -> ConsumerUtils.maybeWrapAsKafkaException(any(Throwable.class))).thenCallRealMethod();
```

Comment:
IntelliJ's editor highlights this line and warns: `Result of 'maybeWrapAsKafkaException()' not thrown`. Suggestions on how to rewrite this welcomed.

Reply:
honestly no idea ;), but well, the whole complexity I see in this test is about mocking this static consumerUtils. Another alternative (simpler) would be to just complete the future of the fetch event with the exception we want to test right? it also removes the thread interrupt code you did not like above (that agree with @AndrewJSchofield , it is what it is). Happy to keep it like this, but if you want to try the alternative I imagine it would be along the lines of :
```
// check timeout
            event1.future().completeExceptionally(new TimeoutException(""));
            assertDoesNotThrow(() -> consumer.poll(Duration.ofMillis(timeoutMs)));
// check interrupt
            event1.future().completeExceptionally(new InterruptException(""));
            assertThrows(reportedFailure.getClass(), () -> consumer.poll(Duration.ofMillis(timeoutMs)));
```
(no  `consumerUtilsMockedStatic` needed at all)

Reply:
@lianetm—thanks for the suggestion.

I was using `Future.completeExceptionally()` at first, but I couldn't get it to work because when `Future.get()` is later invoked by `ConsumerUtils.getResult()`, the exception is always wrapped in an `ExecutionException`. Even the `InterruptedException` and `TimeoutException` were wrapped in the `ExecutionException`, so I wasn't able to test the other `catch` blocks in `ConsumerUtils.getResult()`.

I also attempted to "capture" the `Future` that was enqueued into the `applicationEventQueue` and force its `get()` method to throw my desired exception type, but I wasn't able to get that to work, either 🤷‍♂️

Reply:
What about this (passes locally for me):
```suggestion
        @Test
    public void testOffsetFetchInterrupt() {
        consumer = newConsumer();
        long timeoutMs = 0;
        doReturn(Fetch.empty()).when(fetchCollector).collectFetch(any(FetchBuffer.class));
        consumer.assign(Collections.singleton(new TopicPartition("topic1", 0)));

        consumer.poll(Duration.ofMillis(timeoutMs));
        verify(applicationEventHandler).add(any(FetchCommittedOffsetsEvent.class));
        CompletableApplicationEvent<Map<TopicPartition, OffsetAndMetadata>> event = getLastEnqueuedEvent();
        assertTrue(consumer.hasPendingOffsetFetchEvent());

        event.future().completeExceptionally(new InterruptException(""));
        assertThrows(InterruptException.class, () -> consumer.poll(Duration.ofMillis(timeoutMs)));
        assertTrue(consumer.hasPendingOffsetFetchEvent());
    }
```

Similar for Runtime and Timeout but adjusting the expectations:
```suggestion
        // Runtime 
        event.future().completeExceptionally(new RuntimeException(""));
        assertThrows(KafkaException.class, () -> consumer.poll(Duration.ofMillis(timeoutMs)));
        assertFalse(consumer.hasPendingOffsetFetchEvent());
        
        // Timeout 
        event.future().completeExceptionally(new TimeoutException(""));
        assertDoesNotThrow(() -> consumer.poll(Duration.ofMillis(timeoutMs)));
        assertTrue(consumer.hasPendingOffsetFetchEvent());
```

All 3 passed for me. (Again, suggesting to simplify getting rid of all the static mocking and warnings, but feel free to leave it for follow-up if you prefer)

Reply:
Interestingly, that approach is still going through the first catch block in `ConsumerUtils.getResult()`:

```java
try {
    return future.get(timer.remainingMs(), TimeUnit.MILLISECONDS);
} catch (ExecutionException e) {
    throw maybeWrapAsKafkaException(e.getCause());
}
. . .
```

However, because we use the `getCause()`, we end up getting the underlying exception passed to `completeExceptionally()`. The only exception type we end up wrapping is the `NullPointerException` case, which we do want wrapped in a `KafkaException`.

OK, I'll make that change. It's much prettier 🌹

Reply:
I've made the tests more clear now. Thanks @lianetm!